### PR TITLE
Feature/Programming-Exercises/Run test cases after due date

### PR DESCRIPTION
**Pull request by thilo-behnke**
### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context

Some courses may require test cases, that are only run after the exercise due date has passed.
This way the instructors can hold tests back so that the students are not bruteforcing their way through a submission.

### Description

- Test cases can now have the flag `afterDueDate`
- When the build result comes from our CI, we identify test cases that have this flag to update the resultString (e.g. `1 of 1 passed (preliminary)` (before due date) vs `2 of 4 passed` (after due date)) and delete feedbacks for after due date tests. The score is not changed (would be both `50%` before and after the due date in the example above).

This PR also contains refactoring on the test-case-manage component to make the inline editing table fields reusable and improving the `updateScoreFromTestCases` control flow in the ResultService.

**Known limitations:**
- ~As the result with all test cases will be run _after_ the due date has passed, it will have the "ungraded" flag. I'm not sure how we should handle this.~ => https://github.com/ls1intum/Artemis/pull/712
- With the current implementation, afterDueDate test cases should not be part of the problem statement

- [x] Implement editing mode in client for changing the afterDueDate setting
- [x] Add new field afterDueDate to programming exercise test case
- [x] Refactor some of the ngx table code to make it more reusable (editable fields)
- [x] Implement functionality to update the afterDueDate in the testCaseResource
- [x] Implement result feedback filtering for test cases before the due date passes
- [x] Write tests for generic table edit components
- [x] Write tests for manage-test-cases table
- [x] Write more service tests for excluding afterDueDate test cases from calculation
- [x] Don’t show 100% but correct score, show info in result string “(preliminary)”

### Steps for Testing

1. Log in to ArTEMiS
2. Go to course administration > programming exercises > manage test cases
3. Set the afterDueDate flag for a couple of test cases
4. Give the programming exercise a due date in the future
5. Go to the code editor, submit
6. Check that the afterDueDate test cases are _not_ part of the feedback and result string (should end with "preliminary")
7. Set the programming exercise due date to the past
8. Go to the code editor, submit
9. Check that the afterDueDate test cases are now part of the feedback and result string

### Screenshots

**Set afterDueDate flag:**

image

**Before due date runs out:**

before-due-date

**Full score before due date:**

image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java [hunk 1]
```diff
@@ -148,13 +170,21 @@ public Result updateResultFromTestCases(Result result, ProgrammingExercise exerc
             // Recalculate the achieved score by including the test cases individual weight.
             if (successfulTestCases.size() > 0) {
                 long successfulTestScore = successfulTestCases.stream().map(ProgrammingExerciseTestCase::getWeight).mapToLong(w -> w).sum();
-                long maxTestScore = testCases.stream().map(ProgrammingExerciseTestCase::getWeight).mapToLong(w -> w).sum();
+                long maxTestScore = testCasesForCurrentDate.stream().map(ProgrammingExerciseTestCase::getWeight).mapToLong(w -> w).sum();
```
Comment (ungaralex) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java]: Out of interest: Why is the score a of type long?

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 2]
```diff
@@ -71,9 +71,9 @@
                 "resetWeights": "Gewichtung zurücksetzen",
                 "weightsReset": "Die Gewichtung wurde erfolgreich zurückgesetzt.",
                 "weightsResetFailed": "Die Gewichtung konnte nicht zurückgesetzt werden.",
-                "saveWeights": "Gewichtung speichern",
-                "weightsUpdated": "Die Gewichtung wurde erfolgreich gespeichert.",
-                "weightCouldNotBeUpdated": "Die Gewichtung für den Testfall {{id}} konnte nicht übernommen werden: {{error}}."
+                "saveTestCases": "Testfälle speichern",
+                "testCasesUpdated": "Die Testfälle wurde erfolgreich gespeichert.",
```
Comment (ungaralex) [inline src/main/webapp/i18n/de/programmingExercise.json]: You forgot an "n" at the end of "wurde" ;)

File: src/test/java/de/tum/in/www1/artemis/ProgrammingExerciseTestCaseServiceTest.java [hunk 3]
```diff
@@ -160,4 +162,82 @@ public void shouldRecalculateScoreBasedOnTestCasesWeight() {
         assertThat(result.getScore()).isEqualTo(expectedScore);
     }
 
+    @Test
+    public void shouldRemoveTestsWithAfterDueDateFlagIfDueDateHasNotPassed() {
+        // Set programming exercise due date in future.
+        programmingExercise.setDueDate(ZonedDateTime.parse("2999-06-30T12:30:40Z[UTC]"));
```
Comment (ungaralex) [inline src/test/java/de/tum/in/www1/artemis/ProgrammingExerciseTestCaseServiceTest.java]: Opinion: I think using ZonedDateTime.now().plus.... is more readable and still gives you a date in the future.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java [hunk 4]
```diff
@@ -123,23 +132,36 @@ public boolean generateTestCasesFromFeedbacks(List<Feedback> feedbacks, Programm
...
+     * Updates an incoming result with the information of the exercises test cases. This update includes:
+     * - Checking which test cases were not executed as this is not part of the bamboo build (not all test cases are executed in an exercise with sequential test runs)
+     * - Checking the due date and the afterDueDate flag
+     * - Recalculating the score based based on the successful test cases weight vs the total weight of all test cases.
      *
      * @param result   to modify with new score, result string & added feedbacks (not executed tests)
      * @param exercise the result belongs to.
      * @return
      */
     public Result updateResultFromTestCases(Result result, ProgrammingExercise exercise) {
```
[the remark this answers is not in the document]
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java]: You are right, I split it up into multiple methods.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java (hunk 1 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java]: I think this is the recommendation from jhipster

File: src/main/webapp/i18n/de/programmingExercise.json (hunk 2 above)
Comment (thilo-behnke) [reply src/main/webapp/i18n/de/programmingExercise.json]: Du hast Recht ;)

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java [hunk 5]
```diff
@@ -123,23 +132,36 @@ public boolean generateTestCasesFromFeedbacks(List<Feedback> feedbacks, Programm
...
+        boolean calculateScoresForAfterDueDateTestCases = exercise.getDueDate() == null || ZonedDateTime.now().isAfter(exercise.getDueDate());
+        // Remove all test cases from the score calculation that are only executed after due date if the due date has not yet passed.
         Set<ProgrammingExerciseTestCase> testCases = findActiveByExerciseId(exercise.getId());
+        Set<ProgrammingExerciseTestCase> testCasesForCurrentDate = testCases.stream().filter(testCase -> calculateScoresForAfterDueDateTestCases || !testCase.isAfterDueDate())
```
[the remark this answers is not in the document]
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java]: I would like to, but Spotless formats it this way :/
I think it's good to extract the filter predicates as much as possible so I tried to improve it this way.

File: src/test/java/de/tum/in/www1/artemis/ProgrammingExerciseTestCaseServiceTest.java (hunk 3 above)
Comment (thilo-behnke) [reply src/test/java/de/tum/in/www1/artemis/ProgrammingExerciseTestCaseServiceTest.java]: Thanks! I didn't know that.

**Comment by thilo-behnke** [comment]
@ungaralex 
Thanks for your review! I made changes as requested, especially splitting up the updateResultFromTestCases split up was a good idea :)

I also fixed a bug that I found: The due date filter should not count for template/solution participation build results as they are indicators for the instructor (also the solution results are used for generating test cases).

**Comment by thilo-behnke** [comment]
This feature is back in progress as Stephan and me have discovered some issues with the implementation:
We have to trigger the students participation again after the due date and create a new result so that the new test cases are evaluated.
It would be better if we could evaluate the the student's result with all test cases and then filter some feedbacks until the due date has passed OR we schedule a job to run after the due date passes.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java [hunk 6]
```diff
@@ -125,41 +135,132 @@ public boolean generateTestCasesFromFeedbacks(List<Feedback> feedbacks, Programm
...
      * @param result   to modify with new score, result string & added feedbacks (not executed tests)
      * @param exercise the result belongs to.
-     * @return the updated result
+     * @param isStudentParticipation boolean flag indicating weather the participation of the result is not a solution/template participation.
+     * @return Result with updated feedbacks, score and result string.
      */
-    public Result updateResultFromTestCases(Result result, ProgrammingExercise exercise) {
+    @Transactional
+    public Result updateResultFromTestCases(Result result, ProgrammingExercise exercise, boolean isStudentParticipation) {
+        boolean removeAfterDueDateTests = isStudentParticipation && exercise.getDueDate() != null && ZonedDateTime.now().isBefore(exercise.getDueDate());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java]: removeAfterDueDateTests sounds like a strange variable name

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java [hunk 7]
```diff
@@ -86,19 +86,20 @@ public ProgrammingExerciseTestCaseResource(ProgrammingExerciseTestCaseRepository
...
-     * @param testCaseWeightUpdates of the test cases to update the weights of.
+     * @param testCaseProgrammingExerciseTestCaseDTOS of the test cases to update the weights and afterDueDate flag of.
      * @return
      */
-    @PatchMapping(value = "programming-exercise/{exerciseId}/update-test-case-weights")
+    @PatchMapping(value = "programming-exercise/{exerciseId}/update-test-cases")
     @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
-    public ResponseEntity<Set<ProgrammingExerciseTestCase>> updateWeights(@PathVariable Long exerciseId, @RequestBody Set<WeightUpdate> testCaseWeightUpdates) {
-        log.debug("REST request to update the weights {} of the exercise {}", testCaseWeightUpdates, exerciseId);
+    public ResponseEntity<Set<ProgrammingExerciseTestCase>> updateWeights(@PathVariable Long exerciseId,
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java]: I guess this method name should also be updated to reflect that more than the weights can change

**Comment by krusche** [comment]
By the way: it would be good if @ungaralex can also review the PR

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java (hunk 6 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java]: You are right, I made it more explicit: `shouldTestsWithAfterDueDateFlagBeRemoved`

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java (hunk 7 above)
Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseTestCaseResource.java]: You are right, I didn't notice. I called it now `updateTestCases`.

**Comment by thilo-behnke** [comment]
@krusche The fixes from your review are done :) I agree that another review from @ungaralex would be helpful.

**Pull request merged**